feat(common): detect LCP images in `NgOptimizedImage` and assert if `priority` is set

AndrewKushnir opened the pull request:
This commit adds extra logic into the `NgOptimizedImage` experimental directive to detect an LCP image and assert whether the `priority` attribute is applied.

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
feat(common): detect LCP images in `NgOptimizedImage` and assert if `priority` is set

This commit adds extra logic into the `NgOptimizedImage` experimental directive to detect an LCP image and assert whether the `priority` attribute is applied.

AndrewKushnir commented:
@kara @pkozlowski-opensource I did some manual testing to verify that it works, but I'd need to add some e2e tests as well. I will do that as a part of this PR within the next couple days, but I'd like to ask if you may have some feedback on the main logic implementation. Thank you.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image.ts:
```diff
@@ -94,6 +170,17 @@ export class NgOptimizedImage implements OnInit {
+  set priority(value: string|boolean|undefined) {
+    this._priority = inputToBoolean(value);
+  }
+  get priority(): boolean {
```
I don't see the new `priority` property being used to reflect it value in the DOM. I'm assuming that this is because we intend it to be used `<img priority>`? But what happens if someone writes `<img [priority]="false">`? Does it fall into the category of bindings that we should detect and error on if they change a value?

AndrewKushnir replied:
Sorry, forgot to mention: this is just a placeholder for this input, the logic will be added in https://github.com/angular/angular/pull/45757. For the purposes of this PR, we don't need to have a reflection of this input into the DOM, the `LCPImageObserver` receives the `priority` value from the `NgOptimizedImage` directive directly (without reading it from DOM).

kara commented on packages/common/src/directives/ng_optimized_image.ts:
```diff
@@ -39,6 +40,78 @@ export const IMAGE_LOADER = new InjectionToken<ImageLoader>('ImageLoader', {
+          const directiveDetails = imgDirectiveDetails({rawSrc: img.rawSrc} as any);
+          console.warn(formatRuntimeError(
+              RuntimeErrorCode.LCP_IMG_MISSING_PRIORITY,
+              `${directiveDetails}: the image was detected as the Largest Contentful Paint (LCP). ` +
```
```suggestion
              `${directiveDetails}: the image was detected as the Largest Contentful Paint (LCP) element, so its loading should be prioritized for optimal performance. ` +
```

kara reviewed (approved):
LGTM

khempenius commented:
I think this code checks whether all LCP candidate images are marked `priority`. I think it would be preferable to only check and assert whether the final LCP candidate image is marked as `priority`. We can also add a test that covers this.

As an example, the first image to load on a page, by virtue of being the only thing on the page so far, is often a LCP candidate and gets reported by PerformanceObserver but isn't necessarily the LCP element.

AndrewKushnir commented:
@khempenius thanks for the comment! Could you please let me know which mechanism can we use to distinguish between an LCP candidate and a final LCP element?

AndrewKushnir commented:
Quick update after offline discussion: instead of going through the list of LCP candidates, we need to look at the last element on the list, which would represent the LCP element more accurately.

AndrewKushnir committed:
test(common): add e2e tests for LCP check logic of the `NgOptimizedImage` directive

This commit adds e2e tests for the LCP check logic. Those tests are needed to verify the behavior in a real browser (vs relying on a Node environment).
